tcms/telemetry/api.py

```diff
@@ -169,3 +168,51 @@ def _append_status_counts_to_result(count, result):
     for status in TestExecutionStatus.chart_status_names:
         status_count = count.get(status, 0)
         result.get(status).append(status_count)
+
+
+@rpc_method(name='Testing.test_case_health')
+def test_case_health(query=None):
+
+    if query is None:
+        query = {}
+
+    passed_cases = _count_cases_for(TestExecutionStatus.passed_status_names, query)
+    failed_cases = _count_cases_for(TestExecutionStatus.failure_status_names, query)
```

Comment:
Consider this example:

- first we take the total execution count (filtered by TP to make it shorter):
```
>>> for te in TestExecution.objects.filter(run__plan=2).values('case_id', 'case__summary').annotate(total=Count('case_id')).order_by('case_id'):
...     print(te)
... 
{'case_id': 17, 'case__summary': 'Training: git basics', 'total': 6}
{'case_id': 18, 'case__summary': 'Training: Python Module 01: Structure of a Python program', 'total': 6}
{'case_id': 19, 'case__summary': 'Training: Python Module 04: Classes and objects', 'total': 6}
{'case_id': 20, 'case__summary': 'Training: Python Module 05: Unit testing', 'total': 6}
{'case_id': 21, 'case__summary': 'Training: Python Module 02: Data types and structures', 'total': 6}
{'case_id': 22, 'case__summary': 'Training: Python Module 03: If statements and loops', 'total': 6}
{'case_id': 23, 'case__summary': 'Training: pip install Django', 'total': 6}
{'case_id': 24, 'case__summary': 'Training: Writing your first Django app, part 1', 'total': 6}
{'case_id': 25, 'case__summary': 'Training: Writing your first Django app, part 2', 'total': 6}
{'case_id': 26, 'case__summary': 'Training: Writing your first Django app, part 3', 'total': 6}
{'case_id': 27, 'case__summary': 'Training: Writing your first Django app, part 4', 'total': 6}
{'case_id': 28, 'case__summary': 'Training: Writing your first Django app, part 5', 'total': 6}
{'case_id': 29, 'case__summary': 'Training: Writing your first Django app, part 6', 'total': 6}
{'case_id': 30, 'case__summary': 'Training: Writing your first Django app, part 7', 'total': 6}
{'case_id': 33, 'case__summary': 'On boarding: create account at tcms.kiwitcms.org', 'total': 1}
```


Here are the ones which are passing:

```
>>> for te in TestExecution.objects.filter(run__plan=2, status__name__in=TestExecutionStatus.passed_status_names).values('case_id', 'case__summary').annotate(total=Count('case_id')).order_by('case_id'):
...     print(te)
... 
{'case_id': 17, 'case__summary': 'Training: git basics', 'total': 2}
{'case_id': 18, 'case__summary': 'Training: Python Module 01: Structure of a Python program', 'total': 2}
{'case_id': 19, 'case__summary': 'Training: Python Module 04: Classes and objects', 'total': 2}
{'case_id': 20, 'case__summary': 'Training: Python Module 05: Unit testing', 'total': 1}
{'case_id': 21, 'case__summary': 'Training: Python Module 02: Data types and structures', 'total': 2}
{'case_id': 22, 'case__summary': 'Training: Python Module 03: If statements and loops', 'total': 2}
```

Here are the ones which are failing:

```
>>> for te in TestExecution.objects.filter(run__plan=2, status__name__in=TestExecutionStatus.failure_status_names).values('case_id', 'case__summary').annotate(total=Count('case_id')).order_by('case_id'):
...     print(te)
... 
{'case_id': 17, 'case__summary': 'Training: git basics', 'total': 1}
{'case_id': 20, 'case__summary': 'Training: Python Module 05: Unit testing', 'total': 1}
```

Notice it is not all of them b/c some TE are still IDLE or RUNNING, e.g. not completed yet.

Based on the information above we can conclude that:

1) There are no test cases with 100% failure rate
2) TC-17 is flaky b/c its failing rate is 1/6 == 17% fail vs. 2/6 == 33 % pass. The rest of the percentage is unknown (running, idle, etc) and is not part of the statistics. Once every bit of testing is complete the data will change.

Reply:
yes, some of them are missing, because we filter by `TestExecutionStatus.passed_status_names`, which are `(PASSED, WAIVED)` and `failure_status_names` - `('ERROR', FAILED)`, e.g. any executions that are `IDLE/RUNNING` will not be counted.

I can't understand from the comment - is there something that I need or change or something that I have gotten wrong, or is it just FYI?

Reply:
this is the query you should be using